File: 022/generate-parentheses.go
func generateParenthesis(n int) []string {
	var ans []string

	var dfs func(string, int, int, int)
	dfs = func(s string, left int, right int, n int) {
		if left == n && right == n {
			ans = append(ans, s)
			return
		}

		if left+1 <= n {
			dfs(s+"(", left+1, right, n)
		}

		if right+1 <= n && right+1 <= left {
			dfs(s+")", left, right+1, n)
		}
	}

	dfs("", 0, 0, n)
	return ans
}

File: 022/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        ans = []

        def dfs(s, left, right, n):
            if left == n and right == n:
                ans.append(s)
                return

            if left + 1 <= n:
                dfs(s + '(', left + 1, right, n)

            if right + 1 <= n and right + 1 <= left:
                dfs(s + ')', left, right + 1, n)

        dfs('', 0, 0, n)
        return ans
